### PR TITLE
Implement a property to allow for splitting tests.

### PR DESCRIPTION
**Pull request by alpar-t**
With this change we will allow one to run 
```
./gradlew -Dtests.split=1/2 check
```
and
```
./gradlew -Dtests.split=2/2 check
```

To run tests in separate invocations possibly across separate machines.
The way the tests are split is meant to be opaque and up to the build.
The intent is to use this in CI to speed up PR checks.

### CODE REVIEW

**Comment by atorok** [comment]
@elasticmachine test this please

**Review by DaveCTurner — COMMENTED** [review]
I think it'd be good if one of the conditions was more explicitly "none of the other conditions match". One way to do this would be to have a list of closures (currently length 1, containing only `{ task -> task.onlyIf {task.path.startsWith(":x-pack:") } }`), avoid the `switch`, and compute the currently-hard-coded `2`s from the length of the list.

It's not clear how much we're likely to want to alter/add to these conditions in future so maybe this is overkill.

**Comment by atorok** [comment]
Thanks for the review @DaveCTurner.  
I do like the idea of making "no match"  more explicit, or rather the fact that we do get all the tests, so I  made a change in that direction. 
I think a more generic approach that makes it easy to split multiple ways is not something we want to do now. Making it easy would suggest it's an accepted direction going forward, which this is not, it's to be considered an experiment or temporary solution for now. 
I also think that the overhead right now is high and we would get diminishing returns if we split more ways.

**Comment by DaveCTurner** [comment]
> I think a more generic approach that makes it easy to split multiple ways is not something we want to do now. Making it easy would suggest it's an accepted direction going forward, which this is not, it's to be considered an experiment or temporary solution for now.

+1 to that. I have no further comments, but do not think I know enough about the build to give a meaningful approval, so I'm deferring the final LGTM to other reviewers.

**Comment by atorok** [comment]
@rjernst @nik9000 can you take a look please ?

**Comment by nik9000** [comment]
I think I'd prefer a list of closures just because I think it'd be easier to read and know "this will certainly cover all of the cases".

**Comment by atorok** [comment]
I think it's the other way around.  
Having a list of closures would not guarantee that. 
We would need an explicit way to see that all tasks are matched by at least one closure, whereas with a predicate and predicate.negate() guarantee it by default.

**Review by rjernst — COMMENTED** [review]
Could this be done more simply as actual tasks? I don't like silently ignoring the actual value of the `tests.split` property, when the suggested values imply the ability to arbitrarily split. Instead, if we had `checkPart1` and `checkPart2`, we can always add a `checkPart3` in the future to further granulate the tests.

**Comment by atorok** [comment]
@rjernst I like your proposal and took this stab at implementing it:
```
 Predicate discriminator = { it.path.startsWith(":x-pack:") }

  task checkPart1 {
    group = "check"
    dependsOn tasks.withType(RandomizedTestingTask) { task ->
      task.onlyIf { discriminator.negate().test(it) }
    }
  }

  task checkPart2 {
    group = "check"
    dependesOn tasks.withType(RandomizedTestingTask) { task ->
      task.onlyIf { discriminator.test(it) }
    }
  }
```
The problem with this approach is that we can have tasks other than `RandomizedTestingTask` that are ran as part of `check` but will be missed when ran like this. With the property based approach we can't possibly change the task graph, but only disable tasks from it. 
I'm not sure what you  mean by silently ignoring the property. 
We do take both values into account and enforce that the split is done 2 ways.

**Comment by rjernst** [comment]
I don't think tasks.withType needs to be used at all. Instead, I would create the two tasks in the root project, and have a project.gradle.projectsEvaluated block. It would iterate over `allprojects`, and call `checkPartN.dependsOn` with the `check` tasks of those subprojects. An if/else can check for the current project being rooted by `x-pack` to design which `N` to use.

> I'm not sure what you mean by silently ignoring the property.

This was my mistake. I misread the isEmpty() if condition and thought that was the only read of the property value.

**Comment by atorok** [comment]
@rjernst is this what you had in mind ?

File: build.gradle [hunk 1]
```diff
@@ -622,3 +625,19 @@ allprojects {
     }
   }
 }
+
+allprojects {
+  task checkPart1
+  task checkPart2 
+  tasks.matching {it.name == "check"}.all { task ->
```
Comment (rjernst) [inline build.gradle]: nit: I would not name the local "task" since that masks the keyword of the same name

**Review by rjernst — COMMENTED** [review]
Yes, but doesn't it need to happen in a `projectsEvaluated` block so that all subprojects have been configured? Otherwise it won't find check tasks for any subprojects not already configured, since that is added by the java plugin (through our build plugin) being applied?

**Comment by atorok** [comment]
@rjernst I don't think so. `matching` and all `all` are `NamedDomainObject` methods, so the closures will be called after the tasks are added as well and not just on the existing elements of the collection at the time the method is called.

**Review by rjernst — APPROVED** [review]
Thanks @alpar. You are right. LGTM

**Pull request merged**